src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java

```diff
@@ -496,14 +500,46 @@ private int buildsOfProjectOnNode(Node node, Task task) {
         return runCount;
     }
 
+    @CheckForNull
+    private Job taskToJob(Task task) {
+        if (task instanceof Job) {
+            return (Job) task;
+        } else if (task instanceof PlaceholderTask) {
+            // TODO: Is this branch necessary? I couldn't trigger this code path in my testing, but elsewhere in this
+            // file we check for PlaceholderTask, so perhaps this is needed for some use case I haven't thought of.
+            PlaceholderTask placeholderTask = (PlaceholderTask) task;
+            Run run = placeholderTask.run();
+            if (run != null) {
+                return run.getParent();
+            }
+        }
+
+        LOGGER.log(Level.WARNING, "Could not dereference job from task of type " + task.getClass().getName());
+        return null;
+    }
+
     private int buildsOfProjectOnAllNodes(Task task) {
-        final Jenkins jenkins = Jenkins.getActiveInstance();
-        int totalRunCount = buildsOfProjectOnNode(jenkins, task);
+        if (!shouldBeThrottled(task, getThrottleJobProperty(task))) {
+            return 0;
+        }
 
-        for (Node node : jenkins.getNodes()) {
-            totalRunCount += buildsOfProjectOnNode(node, task);
+        // Prefer Job#getBuilds() to counting executors because the former doesn't count flyweight executors.
+        Job job = taskToJob(task);
+        if (job != null) {
+            RunList<?> runningBuilds = job.getBuilds().filter(new Predicate<Run>() {
```

Comment:
I think this will cause all builds for the job to be loaded from disk, which seems like it could be a bad idea for performance if you have a lot of build history. You could use `FlowExecutionList` to get only running Pipelines and filter that down to the job in question, which has the benefit that all of those builds are guaranteed to already be loaded since they are running. Maybe even keep the old code path for non-Pipeline builds?

Reply:
Great catch! I've done exactly as you suggested, including keeping the old code path for non-Pipeline builds.